arangod/Cluster/ClusterInfo.cpp

```diff
@@ -1504,219 +1538,269 @@ int ClusterInfo::createCollectionCoordinator(
     {
       // check against planned views as well
       AllViews::const_iterator it = _plannedViews.find(databaseName);
       if (it != _plannedViews.end()) {
-        DatabaseViews::const_iterator it2 = (*it).second.find(name);
+        DatabaseViews::const_iterator it2 = (*it).second.find(info.name);
 
         if (it2 != (*it).second.end()) {
           // view already exists!
-          events::CreateView(name, TRI_ERROR_ARANGO_DUPLICATE_NAME);
+          events::CreateCollection(info.name, TRI_ERROR_ARANGO_DUPLICATE_NAME);
           return TRI_ERROR_ARANGO_DUPLICATE_NAME;
         }
       }
     }
-  }
 
-  LOG_TOPIC(DEBUG, Logger::CLUSTER) << "createCollectionCoordinator, checking things...";
+    LOG_TOPIC(DEBUG, Logger::CLUSTER)
+        << "createCollectionCoordinator, checking things...";
 
-  // mop: why do these ask the agency instead of checking cluster info?
-  if (!ac.exists("Plan/Databases/" + databaseName)) {
-    events::CreateCollection(name, TRI_ERROR_ARANGO_DATABASE_NOT_FOUND);
-    return setErrormsg(TRI_ERROR_ARANGO_DATABASE_NOT_FOUND, errorMsg);
-  }
+    // mop: why do these ask the agency instead of checking cluster info?
+    if (!ac.exists("Plan/Databases/" + databaseName)) {
+      events::CreateCollection(info.name, TRI_ERROR_ARANGO_DATABASE_NOT_FOUND);
+      return TRI_ERROR_ARANGO_DATABASE_NOT_FOUND;
+    }
 
-  if (ac.exists("Plan/Collections/" + databaseName + "/" + collectionID)) {
-    events::CreateCollection(name, TRI_ERROR_CLUSTER_COLLECTION_ID_EXISTS);
-    return setErrormsg(TRI_ERROR_CLUSTER_COLLECTION_ID_EXISTS, errorMsg);
+    if (ac.exists("Plan/Collections/" + databaseName + "/" + info.collectionID)) {
+      events::CreateCollection(info.name, TRI_ERROR_CLUSTER_COLLECTION_ID_EXISTS);
+      return TRI_ERROR_CLUSTER_COLLECTION_ID_EXISTS;
+    }
   }
 
   // The following three are used for synchronization between the callback
   // closure and the main thread executing this function. Note that it can
   // happen that the callback is called only after we return from this
   // function!
   auto dbServerResult = std::make_shared<std::atomic<int>>(-1);
+  auto nrDone = std::make_shared<std::atomic<size_t>>(0);
   auto errMsg = std::make_shared<std::string>();
   auto cacheMutex = std::make_shared<Mutex>();
   auto cacheMutexOwner = std::make_shared<std::atomic<std::thread::id>>();
+
+  std::vector<std::shared_ptr<AgencyCallback>> agencyCallbacks;
+
+  auto cbGuard = scopeGuard([&] {
+    for (auto& cb : agencyCallbacks) {
+      _agencyCallbackRegistry->unregisterCallback(cb);
+    }
+  });
+  std::vector<AgencyOperation> opers({IncreaseVersion()});
+
+  std::vector<AgencyPrecondition> precs;
+  std::unordered_set<std::string> conditions;
+
   // current thread owning 'cacheMutex' write lock (workaround for non-recursive Mutex)
+  for (auto& info : infos) {
+    TRI_ASSERT(!info.name.empty());
+    info.dbServerChanged = [cacheMutex, cacheMutexOwner, &info, dbServerResult,
```

Comment:
We have to be very careful here. In the past, we had problems because the agent callback can sometimes come in, when this enclosing function here is already terminated. Therefore everything handed into the closure must be shared_ptr protected. This is the case here, with the exception of &info. This seems to be a problem because this is a local variable in the enclosing function, which is a reference to one entry of the `infos` vector. This vector is passed in from the outside. However, it might already have gone out of scope, when the agency callback comes (or is executed). I do not know the exact circumstances when this happens, but this is one of the subtle bugs which made jenkins runs unstable for a long time.

Reply:
To be on the safe side, we would have to copy *info into the closure...